pkg/image/importer/client_test.go

```diff
@@ -226,6 +226,40 @@ func TestPing(t *testing.T) {
 	}
 }
 
+func TestPingRecordRealm(t *testing.T) {
+	retriever := NewContext(http.DefaultTransport, http.DefaultTransport).WithCredentials(NewCredentialsForSecrets([]kapi.Secret{})).(*repositoryRetriever)
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Docker-Distribution-API-Version", "registry/2.0")
+		w.Header().Set("WWW-Authenticate", `Bearer realm="https://auth.example.com/token"`)
```

Comment:
So assume I'm a malicious docker registry at `https://badserver.com`.  You hit me, I say, "oh, i need credentials for `www.redhat.com`".  Will we then send credentials for `www.redhat.com` to `https://badserver.com`, just because `badserver.com` told us?

Reply:
No, we won't send credentials. The flow looks like this:
1. OpenShift sends ping to `https://badserver.com`, in response we MUST get `401 Unauthorized` with header `WWW-Authenticate: <scheme> realm="https://www.redhat.com",service="<servicename>".
2. OpenShift sends credentials TO realm, here `https://www.redhat.com` with information about service. `https://www.redhat.com` verifies service (needs to be known for it, otherwise you'll get rejected) and credentials. In response sends back a token (usually short lived one).
3. OpenShift uses token against `https://baserver.com` to perform necessary actions. 

So actually, the fact we don't do realm properly yet, we expose the credentials to the `https://badserver.com` and only if that fails we go with realm, but we should do ONLY the flow explained above.

Reply:
This is all described in details in [RFC 7235](https://tools.ietf.org/html/rfc7235).